crates/sim2h_client/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+        let agent_pubkey = HashString::from(enc.encode(&*pub_key).unwrap());
+        info!("Generated agent id: {}", agent_pubkey);
+        let connection = await_in_stream_connect(connect_uri)
+            .map_err(|e| format!("Error awaiting connection: {}", e))?;
+
+        let out = Self {
+            agent_pubkey,
+            pub_key: Arc::new(Mutex::new(pub_key)),
+            sec_key: Arc::new(Mutex::new(sec_key)),
+            connection,
+        };
+
+        Ok(out)
+    }
+
+    pub fn connection(&mut self) -> &mut Connection {
+        &mut self.connection
+    }
+
+    pub fn agent_pubkey(&self) -> AgentPubKey {
+        AgentPubKey::from(self.agent_pubkey.clone())
+    }
+
+    pub fn await_msg<F>(&mut self, predicate: F) -> Result<WireMessage, String>
+    where
+        F: Fn(&WireMessage) -> bool,
+    {
+        let timeout = std::time::Instant::now()
+            .checked_add(std::time::Duration::from_millis(10000))
+            .unwrap();
+
+        loop {
+            if let Some(msg) = pull_message_from_stream(&mut self.connection) {
+                if predicate(&msg) {
+                    return Ok(msg);
+                } else {
+                    println!("await_msg skipping message: {:?}", msg);
+                }
+            }
+
+            if std::time::Instant::now() >= timeout {
+                Err("could not connect within timeout".to_string())?
+            }
+
+            std::thread::sleep(std::time::Duration::from_millis(10));
+        }
+    }
+
+    /// sign a message and send it to sim2h
+    pub fn send_wire(&mut self, message: WireMessage) {
+        let payload: Opaque = message.into();
+        let payload_buf: Box<dyn lib3h_crypto_api::Buffer> = Box::new(payload.clone().as_bytes());
+        let sig = base64::encode(
+            &*CRYPTO
+                .with(|crypto| {
+                    let mut sig = crypto.buf_new_insecure(crypto.sign_bytes());
+                    crypto
+                        .sign(&mut sig, &payload_buf, &*self.sec_key.lock().unwrap())
+                        .unwrap();
+                    sig
+                })
+                .read_lock(),
+        );
+        let signed_message = SignedWireMessage {
+            provenance: Provenance::new(self.agent_pubkey.clone(), sig.into()),
+            payload,
+        };
+        let to_send: Opaque = signed_message.into();
+        self.connection.write(to_send.as_bytes().into()).unwrap();
+    }
+}
+
+fn pull_message_from_stream(connection: &mut Connection) -> Option<WireMessage> {
+    let mut frame = WsFrame::default();
+    match connection.read(&mut frame) {
+        Ok(_) => {
+            if let WsFrame::Binary(b) = frame {
+                let msg: WireMessage = serde_json::from_slice(&b).unwrap();
+                Some(msg)
+            } else {
+                panic!("unexpected {:?}", frame);
+            }
+        }
+        Err(e) if e.would_block() => None,
+        Err(e) => panic!(e),
```

Comment:
This is the line that is panicking.
I think this whole file needs to be matched to sim2h_client.rs in develop